### ISSUE TITLE
Grunticon embed fails on firefox if css contains rules with pseudo elements

### ISSUE THREAD

**Issue by prashantpalikhe** [filamentgroup/grunticon#236]
If the generated svg css contains rules like the following

**icons.data.svg.css**
```
.my-icon:before {
    content: "";
    background-image: url('data:image/svg+xml;charset=US-ASCII...');
    display: block;
    width: 14px;
    height: 14px;
}
```

Grunticon embed fails on firefox as it tries to attempt the following:

**grunticon.embed.js line 64**
```
document.querySelectorAll( '.my-icon::before[data-grunticon-embed]' );
```

which results in 

```
SyntaxError: An invalid or illegal string was specified
```

because the CSS selector given to document.querySelectorAll is invalid since pseudo elements cannot have attributes.

**Comment by jefflembeck** [comment]
Definitely a bug, we just need to decide how to handle this kind of error.

Likely we need to throw an error either way, as shoving an SVG into a `before` isn't a great idea, but should probably better handle how the error occurs.

**Comment by prashantpalikhe** [comment]
I am thinking instead of querying for embed elements _with_ embed attribute, we query only using the selector. Then check if the embed element has the attribute, and proceed only if so.

```javascript
var embedIcons = function(icons){
        var embedElems, embedAttr, selector;

        // attr to specify svg embedding
        embedAttr = "data-grunticon-embed";

        for( var iconName in icons ){
            selector = iconName.slice(selectorPlaceholder.length);
            embedElems = document.querySelectorAll( selector);
            if( !embedElems.length ){ continue; }

            for( var i = 0; i < embedElems.length; i++ ){
                // Proceed only if embed directive set
                if (!embedElems[ i ].hasAttribute(embedAttr)) {
                    continue;
                }

                embedElems[ i ].innerHTML = icons[ iconName ];
                embedElems[ i ].style.backgroundImage = "none";
                embedElems[ i ].removeAttribute( embedAttr );
            }
        }
        return embedElems;
    };
```

Also check the [compatibility](https://developer.mozilla.org/en-US/docs/Web/API/Element/hasAttribute#AutoCompatibilityTable) of Element.hasAttribute()

**Comment by scottjehl** [comment]
Hmm. Running that query on the document in every iteration of the loop seems wasteful. Maybe that could be moved out and cached.

That said, I think we can address the original issue with either try/catch or by filtering out selectors that contain certain pseudo-selectors. I'd vote for the former if we can, since it lets the browser catch these cases for us.

**Comment by prashantpalikhe** [comment]
Well, the number of DOM query remains the same, and since we already have references to the DOM elements in the second loop, the added execution overhead should not be significant.

But try/catch is a good approach indeed.

**Comment by jefflembeck** [comment]
Closed with #238
